addons/sale_mrp/models/account_move.py

```diff
@@ -16,7 +16,11 @@ def _stock_account_get_anglo_saxon_price_unit(self):
             # invoice validation a wrong price will be taken
             boms = so_line.move_ids.filtered(lambda m: m.state != 'cancel').mapped('bom_line_id.bom_id').filtered(lambda b: b.type == 'phantom')
             if boms:
-                bom = boms[0]
+                bom = boms.filtered(lambda b: b.product_id == so_line.product_id or b.product_tmpl_id == so_line.product_id.product_tmpl_id)
+                if not bom:
+                    # In the case where the product has no direct component in its bom, it won't be present in the stock moves boms.
+                    # We then take the first bom of the product.
+                    bom = self.env['mrp.bom']._bom_find(products=so_line.product_id, company_id=so_line.company_id.id, bom_type='phantom')[so_line.product_id]
```

Comment:
```suggestion
                    if len(so_line.product_id.bom_ids) > 1:
                        last_move_date = so_line.move_ids[-1].write_date
                        bom = so_line.product_id.bom_ids.filtered(lambda b: b.write_date <= last_move_date).sorted('write_date', reverse=True)[0]
                    else:
                        bom = so_line.product_id.bom_ids
```

Reply:
This should correct the first point in the comment below, but as I'm no stock expert, I don't know if this can be done like this.